### PR TITLE
Implement compileMatcher

### LINKED ISSUE #74: Implement `MatcherFunction.compileMatcher`
Current state:

```scala

@autoContravariant
trait MatcherFunction[-In] {
  def matches(in: In): Matched[Unit]
}

//...

val compileMatcher: Matcher => MatcherFunction[MergeRequestState] = _ => isSuccessful
```

The `compileMatcher` function ignores every part of the MergeRequestState, except for the pipeline's status.

It should support all kinds of the `Matcher` ADT, including the `Many` node which requires all its children to match in order to match itself. An empty list in `Many` should mean "always match" (in the future we could consider making it a NEL instead of a list).

The rule currently used is the one defined in `ProjectConfigReader.test`: `Rule.mergeAnything`:

`val mergeAnything = Rule("anything", Matcher.Many(Nil), Action.Merge)`

it should probably be renamed to something like `mergeSuccessful` and contain a predicate based on the pipeline's status.

---

The `compileMatcher` function should be relatively easy to test. One set of test cases should definitely use the Scala Steward config defined in `ProjectConfigReader.test` (it can be copy-pasted verbatim). To get sample data, I recommend looking for Scala Steward MRs in the public Gitlab instance (via the graphql API, whose graphiql UI is linked in the readme) or at work ;)

### PR DESCRIPTION
**Pull request by majk-p**
- [x] Implemented `compileMatcher`
- [x] Todos removed
- [x] Tests added

Closes #74.

### CODE REVIEW

File: src/main/scala/io/pg/actions.scala [hunk 1]
```diff
@@ -45,12 +48,20 @@ object ProjectActions {
 
   object MatcherFunction {
 
+    implicit def monoidK: MonoidK[MatcherFunction] = new MonoidK[MatcherFunction] {
```
Comment (kubukoz) [inline src/main/scala/io/pg/actions.scala:51]: This can be a val

File: src/main/scala/io/pg/actions.scala [hunk 2]
```diff
@@ -45,12 +48,20 @@ object ProjectActions {
...
+      override def empty[A]: MatcherFunction[A] = _ => Right(())
+    }
+
     def fromPredicate[In](
       predicate: In => Boolean,
       orElse: In => Mismatch
     ): MatcherFunction[In] =
       _.asRight[Mismatch].ensureOr(orElse)(predicate).toEitherNel.void
 
+    def success[In]: MatcherFunction[In] =
```
Comment (kubukoz) [inline src/main/scala/io/pg/actions.scala:63]: It could be a `MatcherFunction[Any]`

File: src/main/scala/io/pg/actions.scala [hunk 3]
```diff
@@ -64,9 +75,43 @@ object ProjectActions {
...
+        value => Mismatch(s"not successful, actual status: $value")
+      )
+      .contramap(_.description)
+
+  val compileMatcher: Matcher => MatcherFunction[MergeRequestState] = {  
+    case Matcher.Author(email) => autorMatches(email)
+    case Matcher.Description(text) => descriptionMatches(text)
+    case Matcher.PipelineStatus(status) => isSuccessful 
+    case Matcher.Many(Nil) => MatcherFunction.success
+    case state @ Matcher.Many(values) => values.foldMapK(compileMatcher)
```
Comment (kubukoz) [inline src/main/scala/io/pg/actions.scala:113]: `state @` is probably redundant

File: src/main/scala/io/pg/actions.scala [hunk 4]
```diff
@@ -64,9 +75,43 @@ object ProjectActions {
...
+
+  def autorMatches(matcher: TextMatcher): MatcherFunction[MergeRequestState] =
+    MatcherFunction
+      .fromPredicate[Option[String]](
+        x => x.exists(matchTextMatcher(matcher)) ,
+        value => Mismatch(s"not successful, actual status: $value")
+      )
+      .contramap(_.authorEmail)
+
+  // FIXME[MP] both matchers are very similar, could be generic, taking a function (MergeRequestState => Option[String])
```
Comment (kubukoz) [inline src/main/scala/io/pg/actions.scala:99]: I was thinking of something like this, so basically lifting a `MatcherFunction[A]` to `MatcherFunction[Option[A]]`. I also made `matchTextMatcher` return `MatcherFunction[String]`.

```diff
diff --git a/src/main/scala/io/pg/actions.scala b/src/main/scala/io/pg/actions.scala
index fc4c071..0462e8d 100644
--- a/src/main/scala/io/pg/actions.scala
+++ b/src/main/scala/io/pg/actions.scala
@@ -15,6 +15,7 @@ import io.pg.gitlab.Gitlab.MergeRequestInfo
 import io.pg.config.TextMatcher
 import cats.kernel.Monoid
 import cats.MonoidK
+import cats.data.NonEmptyList
 
 @finalAlg
 trait ProjectActions[F[_]] {
@@ -49,7 +50,7 @@ object ProjectActions {
   object MatcherFunction {
 
     implicit def monoidK: MonoidK[MatcherFunction] = new MonoidK[MatcherFunction] {
-      override def combineK[A](x: MatcherFunction[A], y: MatcherFunction[A]): MatcherFunction[A] = 
+      override def combineK[A](x: MatcherFunction[A], y: MatcherFunction[A]): MatcherFunction[A] =
         in => (x.matches(in).toValidated |+| y.matches(in).toValidated).toEither
       override def empty[A]: MatcherFunction[A] = _ => Right(())
     }
@@ -60,7 +61,7 @@ object ProjectActions {
     ): MatcherFunction[In] =
       _.asRight[Mismatch].ensureOr(orElse)(predicate).toEitherNel.void
 
-    def success[In]: MatcherFunction[In] = 
+    def success[In]: MatcherFunction[In] =
       _.pure[Matched].void
   }
 
@@ -75,42 +76,42 @@ object ProjectActions {
       )
       .contramap(_.status)
 
-  def matchTextMatcher(matcher: TextMatcher)(value: String) = {
+  def matchTextMatcher(matcher: TextMatcher): MatcherFunction[String] =
     /*
-    FIXME[MP] 
+    FIXME[MP]
       - move the logic away - perhaps TextMatcher ops?
-      - check if new Regex is safe
-    */
-    import scala.util.matching.Regex 
+      - check if new .r is safe
+     */
     matcher match {
-      case TextMatcher.Equals(expected) => value === expected 
-      case TextMatcher.Matches(regex) => new Regex(regex).matches(value) 
+      case TextMatcher.Equals(expected) =>
+        MatcherFunction.fromPredicate[String](
+          _ === expected,
+          value => Mismatch(show"invalid value, expected $expected, got $value")
+        )
+      case TextMatcher.Matches(regex)   =>
+        MatcherFunction.fromPredicate[String](
+          regex.r.matches,
+          value => Mismatch(show"invalid value, expected to match $regex, got $value")
+        )
     }
-  }
+
+  def exists[A](base: MatcherFunction[A]): MatcherFunction[Option[A]] =
+    _.fold[Matched[Unit]](Mismatch("Option was empty").leftNel)(base.matches)
 
   def autorMatches(matcher: TextMatcher): MatcherFunction[MergeRequestState] =
-    MatcherFunction
-      .fromPredicate[Option[String]](
-        x => x.exists(matchTextMatcher(matcher)) ,
-        value => Mismatch(s"not successful, actual status: $value")
-      )
+    exists(matchTextMatcher(matcher))
       .contramap(_.authorEmail)
 
-  // FIXME[MP] both matchers are very similar, could be generic, taking a function (MergeRequestState => Option[String])
   def descriptionMatches(matcher: TextMatcher): MatcherFunction[MergeRequestState] =
-    MatcherFunction
-      .fromPredicate[Option[String]](
-        x => x.exists(matchTextMatcher(matcher)) ,
-        value => Mismatch(s"not successful, actual status: $value")
-      )
+    exists(matchTextMatcher(matcher))
       .contramap(_.description)
 
-  val compileMatcher: Matcher => MatcherFunction[MergeRequestState] = {  
-    case Matcher.Author(email) => autorMatches(email)
-    case Matcher.Description(text) => descriptionMatches(text)
-    case Matcher.PipelineStatus(status) => isSuccessful 
-    case Matcher.Many(Nil) => MatcherFunction.success
-    case state @ Matcher.Many(values) => values.foldMapK(compileMatcher)
+  val compileMatcher: Matcher => MatcherFunction[MergeRequestState] = {
+    case Matcher.Author(email)          => autorMatches(email)
+    case Matcher.Description(text)      => descriptionMatches(text)
+    case Matcher.PipelineStatus(status) => isSuccessful
+    case Matcher.Many(Nil)              => MatcherFunction.success
+    case state @ Matcher.Many(values)   => values.foldMapK(compileMatcher)
   }
 
   def compile(

```

This doesn't give us too much information (e.g. you won't see which field of the MRState had the mismatch), but we can extend `MatcherFunction` to add more combinators like "named", which would tag every mismatch with a name (or create a group of mismatches for a name) - something we can look at in the future :)

File: src/main/scala/io/pg/actions.scala (hunk 2 above)
Comment (majk-p) [reply src/main/scala/io/pg/actions.scala]: done

File: src/main/scala/io/pg/actions.scala (hunk 1 above)
Comment (majk-p) [reply src/main/scala/io/pg/actions.scala]: done

File: src/main/scala/io/pg/actions.scala (hunk 3 above)
Comment (majk-p) [reply src/main/scala/io/pg/actions.scala]: right, fixed

**Comment by majk-p** [comment]
I've changed the `TextMatcher.Matches` to wrap `Regex`, moving the regex compilation to the single point of loading configuration. @kubukoz please have a look.

File: src/main/scala/io/pg/config/ProjectConfig.scala [hunk 5]
```diff
@@ -27,7 +27,7 @@ object ProjectConfigReader {
         Matcher.Many(
           List(
             Matcher.Author(TextMatcher.Equals("scala-steward@ocado.com")),
-            Matcher.Description(TextMatcher.Matches("*labels:.*semver-patch.*"))
+            Matcher.Description(TextMatcher.Matches(".+labels:.+semver-patch.+".r))
```
Comment (kubukoz) [inline src/main/scala/io/pg/config/ProjectConfig.scala:30]: I think we should have something like we did, but with an extra dot that I apparently missed:

`.*labels:.*semver-patch.*`

Just like in https://github.com/pitgull/pitgull/blob/main/.mergify.yml#L6.

File: src/main/scala/io/pg/config/format.scala [hunk 6]
```diff
@@ -1,23 +1,43 @@
...
 private object circe {
   implicit val circeConfig: Configuration =
     Configuration.default.withDiscriminator("kind")
+
+  private val decodeRegex: Decoder[Regex] = Decoder.instance {
+    _.value
+      .asString
+      .toRight(DecodingFailure("Failed to decode as String", Nil))
+      .flatMap { s =>
+        Either.fromTry(Try(s.r)).leftMap(e => DecodingFailure(s"Failed to compile regex: $e", Nil))
```
Comment (kubukoz) [inline src/main/scala/io/pg/config/format.scala:23]: Either.fromTry -> `Either.catchNonFatal(s.r)`, plus we can use DecodingFailure.fromThrowable to wrap the exception.

**Comment by kubukoz** [comment]
Looks good, now we'll only need some tests :D

File: src/main/scala/io/pg/config/ProjectConfig.scala (hunk 5 above)
Comment (kubukoz) [reply src/main/scala/io/pg/config/ProjectConfig.scala:30]: If you rebase on main, it'll also work with the remote dhall config 😅

File: src/main/scala/io/pg/config/format.scala (hunk 6 above)
Comment (kubukoz) [reply src/main/scala/io/pg/config/format.scala:23]: nitpick^

**Comment by kubukoz** [comment]
Edit: We'll merge without tests, as this project is of disputable future anyway ;)

File: src/main/scala/io/pg/actions.scala [hunk 7]
```diff
@@ -45,12 +47,20 @@ object ProjectActions {
 
   object MatcherFunction {
 
+    implicit val monoidK: MonoidK[MatcherFunction] = new MonoidK[MatcherFunction] {
+      override def combineK[A](x: MatcherFunction[A], y: MatcherFunction[A]): MatcherFunction[A] = 
+        in => (x.matches(in).toValidated |+| y.matches(in).toValidated).toEither
+      override def empty[A]: MatcherFunction[A] = _ => Right(())
```
Comment (kubukoz) [inline src/main/scala/io/pg/actions.scala:53]: This is probably just `= success`

File: src/main/scala/io/pg/actions.scala [hunk 8]
```diff
@@ -64,9 +74,38 @@ object ProjectActions {
...
+
+  def descriptionMatches(matcher: TextMatcher): MatcherFunction[MergeRequestState] =
+    exists(matchTextMatcher(matcher))
+      .contramap(_.description)
+
+  val compileMatcher: Matcher => MatcherFunction[MergeRequestState] = {  
+    case Matcher.Author(email) => autorMatches(email)
+    case Matcher.Description(text) => descriptionMatches(text)
+    case Matcher.PipelineStatus(status) => isSuccessful 
+    case Matcher.Many(Nil) => MatcherFunction.success
```
Comment (kubukoz) [inline src/main/scala/io/pg/actions.scala:106]: We probably don't need this case, as `foldMapK` on an empty list will yield `empty` from the MonoidK instance, which is equivalent to `success`

File: src/main/scala/io/pg/actions.scala [hunk 9]
```diff
@@ -64,9 +74,38 @@ object ProjectActions {
...
+      .contramap(_.authorEmail)
+
+  def descriptionMatches(matcher: TextMatcher): MatcherFunction[MergeRequestState] =
+    exists(matchTextMatcher(matcher))
+      .contramap(_.description)
+
+  val compileMatcher: Matcher => MatcherFunction[MergeRequestState] = {  
+    case Matcher.Author(email) => autorMatches(email)
+    case Matcher.Description(text) => descriptionMatches(text)
+    case Matcher.PipelineStatus(status) => isSuccessful
```
Comment (kubukoz) [inline src/main/scala/io/pg/actions.scala:105]: `isSuccessful` should be changed to take the expected status as a parameter :) + renamed

File: src/main/scala/io/pg/config/ProjectConfig.scala (hunk 5 above)
Comment (majk-p) [reply src/main/scala/io/pg/config/ProjectConfig.scala]: Done

File: src/main/scala/io/pg/config/format.scala (hunk 6 above)
Comment (majk-p) [reply src/main/scala/io/pg/config/format.scala]: done

File: src/main/scala/io/pg/actions.scala (hunk 9 above)
Comment (majk-p) [reply src/main/scala/io/pg/actions.scala]: renamed to `statusMatches` and fixed to work for any expected status

File: src/main/scala/io/pg/actions.scala (hunk 8 above)
Comment (majk-p) [reply src/main/scala/io/pg/actions.scala:117]: Precisely:
```scala
override def empty[A]: MatcherFunction[A] = _ => Right(())
```
I'll change it to `success`, though removing this line implicitly assumes certain behavior of `MatcherFunction`'s monoid instance. There's probably zero chance this monoid would change, but I wonder if it's in general safe to introduce this kind of implicit assumption?

Comment (majk-p) [reply src/main/scala/io/pg/actions.scala]: done

Comment (kubukoz) [reply src/main/scala/io/pg/actions.scala]: We can probably say "empty list is undefined behavior" and call it a day :D  

on a serious note, we should have a test for this - then any assumptions on the instance would be reasonable, I think. We have the assumptions for how `combine` works, so why not for how `empty` works?

File: src/main/scala/io/pg/actions.scala [hunk 10]
```diff
@@ -45,28 +47,67 @@ object ProjectActions {
...
+          case MergeRequestInfo.Status.Other(value) => expectedStatus === value
+        },
+        value => Mismatch(s"""Status is not "$expectedStatus", actual status: "$value"""")
       )
       .contramap(_.status)
 
-  //todo: matching logic :))
-  //let the knife do the work
-  val compileMatcher: Matcher => MatcherFunction[MergeRequestState] = _ => isSuccessful //todo
+  def matchTextMatcher: TextMatcher => MatcherFunction[String] = {
```
Comment (kubukoz) [inline src/main/scala/io/pg/actions.scala:80]:
```suggestion
  val matchTextMatcher: TextMatcher => MatcherFunction[String] = {
```

**Pull request merged**